File: server/privacy-request/application.ts
import {
  planPrivacyRequestCompletion,
  planPrivacyRequestFailure,
  planPrivacyRequestProcessingStart,
  planPrivacyRequestStart,
  planPrivacyRequestVerification,
} from './core';
import {
  privacyRequestPublicStatus,
  type PrivacyRequestPublicStatus,
  type PrivacyRequestSubmitCommand,
} from './application-core';
import {
  parsePrivacyRequestFailureCode,
  privacyRequestScope,
  type PrivacyRequestFailureCode,
  type PrivacyRequestId,
  type PrivacyRequestRecord,
  type PrivacyRequestRepository,
  type PrivacyRequestScope,
  type PrivacyRequestVerificationReceiptId,
} from './public';

export type PrivacyRequestVerificationPortResult =
  | {
      readonly kind: 'approved';
      readonly receiptId: PrivacyRequestVerificationReceiptId;
    }
  | {
      readonly kind: 'rejected';
      readonly reason: 'identity-not-verified' | 'request-not-applicable';
    }
  | { readonly kind: 'unavailable' };

export type PrivacyRequestVerificationPort = {
  verify(input: {
    readonly scope: PrivacyRequestScope;
    readonly requestId: PrivacyRequestId;
    readonly requestKind: PrivacyRequestRecord['requestKind'];
    readonly checkedAt: number;
  }): Promise<PrivacyRequestVerificationPortResult>;
};

export type PrivacyRequestExecutionResult =
  | { readonly kind: 'fulfilled' }
  | {
      readonly kind: 'failed';
      readonly failureCode: PrivacyRequestFailureCode;
      readonly retryable: boolean;
    };

export type PrivacyRequestExecutionPort = {
  execute(input: {
    readonly scope: PrivacyRequestScope;
    readonly requestId: PrivacyRequestId;
    readonly requestKind: Exclude<
      PrivacyRequestRecord['requestKind'],
      'deletion'
    >;
  }): Promise<PrivacyRequestExecutionResult>;
};

export type PrivacyRequestDeletionHandoffResult =
  | { readonly kind: 'started' }
  | {
      readonly kind: 'failed';
      readonly failureCode: PrivacyRequestFailureCode;
      readonly retryable: boolean;
    };

export type PrivacyRequestDeletionHandoffPort = {
  startExistingAccountDeletionSaga(input: {
    readonly scope: PrivacyRequestScope;
    readonly privacyRequestId: PrivacyRequestId;
  }): Promise<PrivacyRequestDeletionHandoffResult>;
};

export type PrivacyRequestApplicationResult =
  | {
      readonly kind: 'accepted';
      readonly outcome: 'recorded' | 'replayed' | 'status' | 'updated';
      readonly request: PrivacyRequestPublicStatus;
    }
  | {
      readonly kind: 'rejected';
      readonly reason:
        | 'identifier-conflict'
        | 'invalid-input'
        | 'invalid-state'
        | 'not-found'
        | 'unavailable';
    };

export type PrivacyRequestApplication = {
  submit(input: {
    readonly scope: PrivacyRequestScope;
    readonly command: PrivacyRequestSubmitCommand;
    readonly requestId: PrivacyRequestId;
    readonly requestedAt: number;
  }): Promise<PrivacyRequestApplicationResult>;
  status(input: {
    readonly scope: PrivacyRequestScope;
    readonly requestId: PrivacyRequestId;
  }): Promise<PrivacyRequestApplicationResult>;
  verify(input: {
    readonly scope: PrivacyRequestScope;
    readonly requestId: PrivacyRequestId;
    readonly checkedAt: number;
  }): Promise<PrivacyRequestApplicationResult>;
  process(input: {
    readonly scope: PrivacyRequestScope;
    readonly requestId: PrivacyRequestId;
    readonly startedAt: number;
    readonly finishedAt: number;
  }): Promise<PrivacyRequestApplicationResult>;
};

export type PrivacyRequestApplicationDependencies = {
  readonly repository: PrivacyRequestRepository;
  readonly verification: PrivacyRequestVerificationPort;
  readonly execution: PrivacyRequestExecutionPort;
  readonly accountDeletion: PrivacyRequestDeletionHandoffPort;
};

const executorUnavailable = parsePrivacyRequestFailureCode(
  'executor-unavailable',
);

export function createPrivacyRequestApplication(
  dependencies: PrivacyRequestApplicationDependencies,
): PrivacyRequestApplication {
  return {
    async submit(input) {
      try {
        const scope = privacyRequestScope(input.scope);
        const plan = planPrivacyRequestStart({
          scope,
          requestId: input.requestId,
          submissionId: input.command.submissionId,
          requestKind: input.command.requestKind,
          requestedAt: input.requestedAt,
        });
        if (plan.kind === 'rejected') return rejected('invalid-input');
        const created = await dependencies.repository.create(plan.record);
        switch (created.kind) {
          case 'created':
            return accepted('recorded', created.record);
          case 'existing':
            return accepted('replayed', created.record);
          case 'conflict':
            return rejected('identifier-conflict');
          case 'rejected':
            return rejected('invalid-input');
        }
      } catch {
        return rejected('unavailable');
      }
    },

    async status(input) {
      try {
        const scope = privacyRequestScope(input.scope);
        const record = await dependencies.repository.findById(
          scope,
          input.requestId,
        );
        return record === undefined
          ? rejected('not-found')
          : accepted('status', record);
      } catch {
        return rejected('unavailable');
      }
    },

    async verify(input) {
      try {
        const scope = privacyRequestScope(input.scope);
        const record = await dependencies.repository.findById(
          scope,
          input.requestId,
        );
        if (record === undefined) return rejected('not-found');
        if (record.state.kind !== 'verification-pending') {
          return accepted('status', record);
        }
        const verification = await dependencies.verification.verify({
          scope,
          requestId: record.requestId,
          requestKind: record.requestKind,
          checkedAt: input.checkedAt,
        });
        if (verification.kind === 'unavailable') {
          return rejected('unavailable');
        }
        const plan = planPrivacyRequestVerification({
          record,
          decision:
            verification.kind === 'approved'
              ? {
                  kind: 'approved',
                  receiptId: verification.receiptId,
                  decidedAt: input.checkedAt,
                }
              : {
                  kind: 'rejected',
                  reason: verification.reason,
                  decidedAt: input.checkedAt,
                },
        });
        return plan.kind === 'accepted'
          ? commitUpdate(dependencies.repository, scope, plan.transition)
          : rejected('invalid-state');
      } catch {
        return rejected('unavailable');
      }
    },

    async process(input) {
      try {
        const scope = privacyRequestScope(input.scope);
        const record = await dependencies.repository.findById(
          scope,
          input.requestId,
        );
        if (record === undefined) return rejected('not-found');
        if (record.state.kind !== 'ready') return accepted('status', record);

        const claim = planPrivacyRequestProcessingStart({
          record,
          startedAt: input.startedAt,
        });
        if (claim.kind === 'rejected') return rejected('invalid-state');
        const claimed = await dependencies.repository.commit(
          scope,
          claim.transition,
        );
        if (claimed.kind === 'rejected') return rejected('invalid-state');
        if (claimed.kind === 'conflict') {
          return claimed.current === undefined
            ? rejected('not-found')
            : accepted('status', claimed.current);
        }
        if (claimed.kind === 'replayed') {
          return accepted('status', claimed.record);
        }

        const execution = await executeRequest(
          dependencies,
          scope,
          claimed.record,
        );
        const completion =
          execution.kind === 'completed'
            ? planPrivacyRequestCompletion({
                record: claimed.record,
                completedAt: input.finishedAt,
                outcome: execution.outcome,
              })
            : planPrivacyRequestFailure({
                record: claimed.record,
                failedAt: input.finishedAt,
                failureCode: execution.failureCode,
                retryable: execution.retryable,
              });
        return completion.kind === 'accepted'
          ? commitUpdate(dependencies.repository, scope, completion.transition)
          : rejected('invalid-state');
      } catch {
        return rejected('unavailable');
      }
    },
  };
}

type RequestExecution =
  | {
      readonly kind: 'completed';
      readonly outcome: 'fulfilled' | 'account-deletion-started';
    }
  | {
      readonly kind: 'failed';
      readonly failureCode: PrivacyRequestFailureCode;
      readonly retryable: boolean;
    };

async function executeRequest(
  dependencies: PrivacyRequestApplicationDependencies,
  scope: PrivacyRequestScope,
  record: PrivacyRequestRecord,
): Promise<RequestExecution> {
  try {
    if (record.requestKind === 'deletion') {
      const result =
        await dependencies.accountDeletion.startExistingAccountDeletionSaga({
          scope,
          privacyRequestId: record.requestId,
        });
      return result.kind === 'started'
        ? { kind: 'completed', outcome: 'account-deletion-started' }
        : result;
    }
    const result = await dependencies.execution.execute({
      scope,
      requestId: record.requestId,
      requestKind: record.requestKind,
    });
    return result.kind === 'fulfilled'
      ? { kind: 'completed', outcome: 'fulfilled' }
      : result;
  } catch {
    return {
      kind: 'failed',
      failureCode: executorUnavailable,
      retryable: true,
    };
  }
}

async function commitUpdate(
  repository: PrivacyRequestRepository,
  scope: PrivacyRequestScope,
  transition: Parameters<PrivacyRequestRepository['commit']>[1],
): Promise<PrivacyRequestApplicationResult> {
  const committed = await repository.commit(scope, transition);
  switch (committed.kind) {
    case 'applied':
    case 'replayed':
      return accepted('updated', committed.record);
    case 'conflict':
      return committed.current === undefined
        ? rejected('not-found')
        : accepted('status', committed.current);
    case 'rejected':
      return rejected('invalid-state');
  }
}

function accepted(
  outcome: Extract<
    PrivacyRequestApplicationResult,
    { kind: 'accepted' }
  >['outcome'],
  record: PrivacyRequestRecord,
): PrivacyRequestApplicationResult {
  return {
    kind: 'accepted',
    outcome,
    request: privacyRequestPublicStatus(record),
  };
}

function rejected(
  reason: Extract<
    PrivacyRequestApplicationResult,
    { kind: 'rejected' }
  >['reason'],
): PrivacyRequestApplicationResult {
  return { kind: 'rejected', reason };
}

File: backend/internal/privacyrequest/application.go
package privacyrequest

import (
	"context"
	"errors"
)

type VerificationResultKind string

const (
	VerificationResultApproved    VerificationResultKind = "approved"
	VerificationResultRejected    VerificationResultKind = "rejected"
	VerificationResultUnavailable VerificationResultKind = "unavailable"
)

type VerificationResult struct {
	Kind      VerificationResultKind
	ReceiptID VerificationReceiptID
	Reason    RejectionReason
}

type VerificationPort interface {
	Verify(context.Context, Scope, RequestID, RequestKind, int64) (VerificationResult, error)
}

type ExecutionResultKind string

const (
	ExecutionFulfilled ExecutionResultKind = "fulfilled"
	ExecutionFailed    ExecutionResultKind = "failed"
)

type ExecutionResult struct {
	Kind        ExecutionResultKind
	FailureCode FailureCode
	Retryable   bool
}

type ExecutionPort interface {
	Execute(context.Context, Scope, RequestID, RequestKind) (ExecutionResult, error)
}

type DeletionHandoffResultKind string

const (
	DeletionHandoffStarted DeletionHandoffResultKind = "started"
	DeletionHandoffFailed  DeletionHandoffResultKind = "failed"
)

type DeletionHandoffResult struct {
	Kind        DeletionHandoffResultKind
	FailureCode FailureCode
	Retryable   bool
}

type DeletionHandoffPort interface {
	StartExistingAccountDeletion(context.Context, Scope, RequestID) (DeletionHandoffResult, error)
}

type ApplicationResultKind string
type ApplicationOutcome string
type ApplicationReason string

const (
	ApplicationAccepted ApplicationResultKind = "accepted"
	ApplicationRejected ApplicationResultKind = "rejected"

	ApplicationRecorded ApplicationOutcome = "recorded"
	ApplicationReplayed ApplicationOutcome = "replayed"
	ApplicationStatus   ApplicationOutcome = "status"
	ApplicationUpdated  ApplicationOutcome = "updated"

	ApplicationIdentifierConflict ApplicationReason = "identifier-conflict"
	ApplicationInvalidInput       ApplicationReason = "invalid-input"
	ApplicationInvalidState       ApplicationReason = "invalid-state"
	ApplicationNotFound           ApplicationReason = "not-found"
	ApplicationUnavailable        ApplicationReason = "unavailable"
)

type ApplicationResult struct {
	Kind    ApplicationResultKind
	Outcome ApplicationOutcome
	Reason  ApplicationReason
	Request *PublicStatus
}

type Service struct {
	repository      Repository
	verification    VerificationPort
	execution       ExecutionPort
	accountDeletion DeletionHandoffPort
}

func NewService(repository Repository, verification VerificationPort, execution ExecutionPort, accountDeletion DeletionHandoffPort) (*Service, error) {
	if repository == nil || verification == nil || execution == nil || accountDeletion == nil {
		return nil, errors.New("complete privacy request dependencies are required")
	}
	return &Service{
		repository: repository, verification: verification,
		execution: execution, accountDeletion: accountDeletion,
	}, nil
}

func (service *Service) Submit(ctx context.Context, scope Scope, command SubmitCommand, requestID RequestID, requestedAt int64) (ApplicationResult, error) {
	if service == nil || service.repository == nil {
		return ApplicationResult{Kind: ApplicationRejected, Reason: ApplicationUnavailable}, nil
	}
	plan := PlanStart(scope, requestID, command.SubmissionID, command.RequestKind, requestedAt)
	if plan.Kind != PlanAccepted {
		return rejectedResult(ApplicationInvalidInput), nil
	}
	created, err := service.repository.Create(ctx, plan.Record)
	if err != nil {
		return ApplicationResult{}, err
	}
	switch created.Kind {
	case CreateCreated:
		return acceptedResult(ApplicationRecorded, created.Record)
	case CreateExisting:
		if created.Record == nil || created.Record.RequestKind != command.RequestKind {
			return rejectedResult(ApplicationIdentifierConflict), nil
		}
		return acceptedResult(ApplicationReplayed, created.Record)
	case CreateConflict:
		return rejectedResult(ApplicationIdentifierConflict), nil
	case CreateRejected:
		return rejectedResult(ApplicationInvalidInput), nil
	default:
		return rejectedResult(ApplicationUnavailable), nil
	}
}

func (service *Service) Status(ctx context.Context, scope Scope, requestID RequestID) (ApplicationResult, error) {
	record, err := service.repository.FindByID(ctx, scope, requestID)
	if err != nil {
		return ApplicationResult{}, err
	}
	if record == nil {
		return rejectedResult(ApplicationNotFound), nil
	}
	return acceptedResult(ApplicationStatus, record)
}

func (service *Service) Verify(ctx context.Context, scope Scope, requestID RequestID, checkedAt int64) (ApplicationResult, error) {
	record, err := service.repository.FindByID(ctx, scope, requestID)
	if err != nil {
		return ApplicationResult{}, err
	}
	if record == nil {
		return rejectedResult(ApplicationNotFound), nil
	}
	if _, pending := record.State.(VerificationPending); !pending {
		return acceptedResult(ApplicationStatus, record)
	}
	verification, err := service.verification.Verify(ctx, scope, requestID, record.RequestKind, checkedAt)
	if err != nil {
		return ApplicationResult{}, err
	}
	if verification.Kind == VerificationResultUnavailable {
		return rejectedResult(ApplicationUnavailable), nil
	}
	decision := VerificationDecision{DecidedAt: checkedAt}
	switch verification.Kind {
	case VerificationResultApproved:
		decision.Kind = VerificationApproved
		decision.ReceiptID = verification.ReceiptID
	case VerificationResultRejected:
		decision.Kind = VerificationRejected
		decision.Reason = verification.Reason
	default:
		return rejectedResult(ApplicationUnavailable), nil
	}
	plan := PlanVerification(*record, decision)
	if plan.Kind != PlanAccepted {
		return rejectedResult(ApplicationInvalidState), nil
	}
	return service.commit(ctx, scope, plan.Transition)
}

func (service *Service) Process(ctx context.Context, scope Scope, requestID RequestID, startedAt int64, finishedAt int64) (ApplicationResult, error) {
	record, err := service.repository.FindByID(ctx, scope, requestID)
	if err != nil {
		return ApplicationResult{}, err
	}
	if record == nil {
		return rejectedResult(ApplicationNotFound), nil
	}
	if _, ready := record.State.(Ready); !ready {
		return acceptedResult(ApplicationStatus, record)
	}
	claim := PlanProcessingStart(*record, startedAt)
	if claim.Kind != PlanAccepted {
		return rejectedResult(ApplicationInvalidState), nil
	}
	claimed, err := service.repository.Commit(ctx, scope, claim.Transition)
	if err != nil {
		return ApplicationResult{}, err
	}
	if claimed.Kind != CommitApplied {
		return resultFromCommit(claimed)
	}
	if claimed.Record == nil {
		return rejectedResult(ApplicationUnavailable), nil
	}
	execution := service.execute(ctx, scope, *claimed.Record)
	var completion Plan
	if execution.Kind == executionCompleted {
		completion = PlanCompletion(*claimed.Record, finishedAt, execution.Outcome)
	} else {
		completion = PlanFailure(*claimed.Record, finishedAt, execution.FailureCode, execution.Retryable)
	}
	if completion.Kind != PlanAccepted {
		return rejectedResult(ApplicationInvalidState), nil
	}
	return service.commit(ctx, scope, completion.Transition)
}

type executionDecisionKind string

const (
	executionCompleted executionDecisionKind = "completed"
	executionFailed    executionDecisionKind = "failed"
)

type executionDecision struct {
	Kind        executionDecisionKind
	Outcome     Outcome
	FailureCode FailureCode
	Retryable   bool
}

func (service *Service) execute(ctx context.Context, scope Scope, record Record) executionDecision {
	unavailable, _ := ParseFailureCode("executor-unavailable")
	if record.RequestKind == KindDeletion {
		result, err := service.accountDeletion.StartExistingAccountDeletion(ctx, scope, record.RequestID)
		if err != nil {
			return executionDecision{Kind: executionFailed, FailureCode: unavailable, Retryable: true}
		}
		if result.Kind == DeletionHandoffStarted {
			return executionDecision{Kind: executionCompleted, Outcome: OutcomeAccountDeletionStarted}
		}
		if result.Kind == DeletionHandoffFailed {
			return executionDecision{Kind: executionFailed, FailureCode: result.FailureCode, Retryable: result.Retryable}
		}
		return executionDecision{Kind: executionFailed, FailureCode: unavailable, Retryable: true}
	}
	result, err := service.execution.Execute(ctx, scope, record.RequestID, record.RequestKind)
	if err != nil {
		return executionDecision{Kind: executionFailed, FailureCode: unavailable, Retryable: true}
	}
	if result.Kind == ExecutionFulfilled {
		return executionDecision{Kind: executionCompleted, Outcome: OutcomeFulfilled}
	}
	if result.Kind == ExecutionFailed {
		return executionDecision{Kind: executionFailed, FailureCode: result.FailureCode, Retryable: result.Retryable}
	}
	return executionDecision{Kind: executionFailed, FailureCode: unavailable, Retryable: true}
}

func (service *Service) commit(ctx context.Context, scope Scope, transition Transition) (ApplicationResult, error) {
	result, err := service.repository.Commit(ctx, scope, transition)
	if err != nil {
		return ApplicationResult{}, err
	}
	return resultFromCommit(result)
}

func resultFromCommit(result CommitResult) (ApplicationResult, error) {
	switch result.Kind {
	case CommitApplied, CommitReplayed:
		return acceptedResult(ApplicationUpdated, result.Record)
	case CommitConflict:
		if result.Current == nil {
			return rejectedResult(ApplicationNotFound), nil
		}
		return acceptedResult(ApplicationStatus, result.Current)
	case CommitRejected:
		return rejectedResult(ApplicationInvalidState), nil
	default:
		return rejectedResult(ApplicationUnavailable), nil
	}
}

func acceptedResult(outcome ApplicationOutcome, record *Record) (ApplicationResult, error) {
	if record == nil {
		return rejectedResult(ApplicationUnavailable), nil
	}
	status, err := PublicStatusFromRecord(*record)
	if err != nil {
		return ApplicationResult{}, err
	}
	return ApplicationResult{Kind: ApplicationAccepted, Outcome: outcome, Request: &status}, nil
}

func rejectedResult(reason ApplicationReason) ApplicationResult {
	return ApplicationResult{Kind: ApplicationRejected, Reason: reason}
}
